File: sumOfPair.js
/*
Given a sorted array and a number x, find a pair in an array whose sum is closest to x.

Input: arr[] = {10, 22, 28, 29, 30, 40}, x = 54
Output: 22 and 30

Input: arr[] = {1, 3, 4, 7, 10}, x = 15
Output: 4 and 10
*/


function sumOfPair(array, x) {
    const arrayLength = array.length
    let minimum = Infinity;
    let index1 = 0, index2 = 0;
    let leftPointer = 0, rightPointer = arrayLength -1;

    while (leftPointer < rightPointer) {
        currentPoinerSum = array[leftPointer] + array[rightPointer]
        difference = Math.abs(currentPoinerSum - x);
        if (difference < minimum) {
            minimum = difference;
            index1 = leftPointer
            index2 = rightPointer
        }g
        if (currentPoinerSum > x) {
            rightPointer -= 1;
        } else {
            leftPointer +=1;
        }
    }
    
    return [ array[index1], array[index2] ]
}

// const inputArray = [10, 22, 28, 29, 30, 40], target = 54;  // --> [ 22, 30 ]
const inputArray = [1, 3, 4, 7, 10], target = 15; // --> [ 4, 10 ]
console.log(sumOfPair(inputArray, target))

File: sum_of_pair.py
"""
Given a sorted array and a number x, find a pair in an array whose sum is closest to x.

Input: arr[] = {10, 22, 28, 29, 30, 40}, x = 54
Output: 22 and 30

Input: arr[] = {1, 3, 4, 7, 10}, x = 15
Output: 4 and 10
"""


def sum_of_pair(arr, x):
    array_length = len(arr)
    minimum = float("inf")
    index_1 = index_2 = 0
    left_index, right_index = 0, array_length -1
    while left_index < right_index:
        differnce = abs(x - (arr[left_index] + arr[right_index]))
        if differnce < minimum:
            index_1 = left_index
            index_2 = right_index
            minimum = differnce

        if arr[left_index] + arr[right_index] > x:
            right_index -= 1
        else:
            left_index +=1
    return [arr[index_1], arr[index_2]]

input_array = [10, 22, 28, 29, 30, 40]; target = 54
# input_array = [1, 3, 4, 7, 10]; target = 15

print(sum_of_pair(input_array, target))